giraph-core/src/main/java/org/apache/giraph/writable/kryo/GiraphClassResolver.java

```diff
@@ -0,0 +1,308 @@
+              name.substring(name.length() - SEQUENCE_NUMBER_LENGTH));
+
+      if (MIN_CLASS_ID == -1) {
+        MIN_CLASS_ID = classId;
+      }
+
+      int adjustedId = classId - MIN_CLASS_ID + BASE_CLASS_ID;
+      if (CLASS_NAME_TO_ID.putIfAbsent(className, adjustedId) == null) {
+        ID_TO_CLASS_NAME.put(adjustedId, className);
+      }
+    }
+  }
+
+  /**
+   * Gets ID for the given class name.
+   * @param className Class name
+   * @return class id Class ID
+   */
+  public static int getClassId(String className) {
+    if (CLASS_NAME_TO_ID.containsKey(className)) {
+      return CLASS_NAME_TO_ID.get(className);
+    }
+    synchronized (GiraphClassResolver.class) {
+      if (CLASS_NAME_TO_ID.containsKey(className)) {
+        return CLASS_NAME_TO_ID.get(className);
+      }
+      refreshCache();
+
+      if (!CLASS_NAME_TO_ID.containsKey(className)) {
+        createClassName(className);
+        refreshCache();
+      }
+    }
+
+    if (!CLASS_NAME_TO_ID.containsKey(className)) {
+      throw new IllegalStateException("Failed to assigned id to " + className);
+    }
+
+    return CLASS_NAME_TO_ID.get(className);
+  }
+
+  /**
+   * Get class name for given ID.
+   * @param id class ID
+   * @return class name
+   */
+  public static String getClassName(int id) {
+    if (ID_TO_CLASS_NAME.containsKey(id)) {
+      return ID_TO_CLASS_NAME.get(id);
+    }
+    synchronized (GiraphClassResolver.class) {
+      if (ID_TO_CLASS_NAME.containsKey(id)) {
+        return ID_TO_CLASS_NAME.get(id);
+      }
+      refreshCache();
+    }
+
+    if (!ID_TO_CLASS_NAME.containsKey(id)) {
+      throw new IllegalStateException("ID " + id + " doesn't exist");
+    }
+    return ID_TO_CLASS_NAME.get(id);
+  }
+
+  @Override
+  public Registration register(Registration registration) {
+    if (registration == null) {
+      throw new IllegalArgumentException("registration cannot be null");
+    }
+    if (registration.getId() == NAME) {
+      throw new IllegalArgumentException("Invalid registration ID");
+    }
+
+    idToRegistration.put(registration.getId(), registration);
+    classToRegistration.put(registration.getType(), registration);
+    if (registration.getType().isPrimitive()) {
+      classToRegistration.put(getWrapperClass(registration.getType()),
+              registration);
+    }
+    return registration;
+  }
+
+  @Override
+  public Registration registerImplicit(Class type) {
+    return register(new Registration(type, kryo.getDefaultSerializer(type),
+            getClassId(type.getName())));
+  }
+
+  @Override
+  public Registration writeClass(Output output, Class type) {
+    if (type == null) {
+      output.writeVarInt(Kryo.NULL, true);
+      return null;
+    }
+
+    Registration registration = kryo.getRegistration(type);
+    if (registration.getId() == NAME) {
+      throw new IllegalStateException("Invalid registration ID");
+    } else {
+      output.writeVarInt(registration.getId() + 2, true);
```

Comment:
Class ID's are incremented by 2 when writing, because 0 is used for null, and 1 is used for non-explicitly registered classes, though non-explicit class registration is an error for this implementation.

Reply:
Can we add this as a comment here?

Reply:
Ok.

Reply:
Ok.